module/os/linux/zfs/zvol_os.c

```diff
@@ -1266,6 +1266,8 @@ zvol_alloc_non_blk_mq(struct zvol_state_os *zso, zvol_queue_limits_t *limits)
 	zso->zvo_queue = zso->zvo_disk->queue;
 
 #ifndef HAVE_BLKDEV_QUEUE_LIMITS_FEATURES
```

Comment:
I don't quite understand why I didn't just call `zvol_queue_limits_apply(limits, zso->zvo_queue)` here, since that was always the point of convert/apply: everything in ~one~ two places.

Reply:
Just curious, why set it in two places if the limits should already be set during initialization?

Reply:
The idea was that you would call `_convert` if you had a creation function that could take limits, and regardless, calling `_apply` afterward in all cases would get everything left over.

(see the comment: I assert in the last paragraph that its "called on all versions").

Oh yeah, this is all in fact redundant. `zvol_alloc_non_blk_mq()` calls `zvol_queue_limits_apply()` at the very end for all cases, so this extra stuff in the middle can go.

```diff
diff --git a/module/os/linux/zfs/zvol_os.c b/module/os/linux/zfs/zvol_os.c
index e04f64e23..122fda3ea 100644
--- module/os/linux/zfs/zvol_os.c
+++ module/os/linux/zfs/zvol_os.c
@@ -1251,7 +1251,6 @@ zvol_alloc_non_blk_mq(struct zvol_state_os *zso, zvol_queue_limits_t *limits)
 
 	zso->zvo_disk->minors = ZVOL_MINORS;
 	zso->zvo_queue = zso->zvo_disk->queue;
-	zvol_queue_limits_apply(limits, zso->zvo_queue);
 #elif defined(HAVE_BLK_ALLOC_DISK_2ARG)
 	struct queue_limits qlimits;
 	zvol_queue_limits_convert(limits, &qlimits);
@@ -1264,11 +1263,6 @@ zvol_alloc_non_blk_mq(struct zvol_state_os *zso, zvol_queue_limits_t *limits)
 	zso->zvo_disk = disk;
 	zso->zvo_disk->minors = ZVOL_MINORS;
 	zso->zvo_queue = zso->zvo_disk->queue;
-
-#ifndef HAVE_BLKDEV_QUEUE_LIMITS_FEATURES
-	blk_queue_set_write_cache(zso->zvo_queue, B_TRUE);
-#endif
-
 #else
 	zso->zvo_queue = blk_alloc_queue(NUMA_NO_NODE);
 	if (zso->zvo_queue == NULL)
```

Reply:
@robn - Supporting different kernel versions has indeed made things quite complex. Hats off to you for managing it so efficiently. I will clean up as per your suggestion along with fixing the original issue.

Reply:
Oh sorry! I was just muttering out loud about my mistake; not actually suggesting you fix it!

(but if you want to anyway, I will be delighted and enthusiastically approve the PR :raised_hands:)